apps/studio/src/lib/db/clients/postgresql.ts

```diff
@@ -39,1347 +38,1419 @@ const pgErrors = {
-  rc.unique_constraint_schema AS to_schema,
-  tc.constraint_name,
-  rc.update_rule,
-  rc.delete_rule,
-  (
-      SELECT STRING_AGG(kcu2.column_name, ',' ORDER BY kcu2.ordinal_position)
-      FROM information_schema.key_column_usage AS kcu2
-      WHERE kcu2.constraint_name = rc.unique_constraint_name
-  ) AS to_column,
-  (
-      SELECT kcu2.table_name
-      FROM information_schema.key_column_usage AS kcu2
-      WHERE kcu2.constraint_name = rc.unique_constraint_name LIMIT 1
-  ) AS to_table
-FROM
-  information_schema.key_column_usage AS kcu
-JOIN
-  information_schema.table_constraints AS tc
-ON
-  tc.constraint_name = kcu.constraint_name
-JOIN
-  information_schema.referential_constraints AS rc
-ON
-  rc.constraint_name = kcu.constraint_name
-WHERE
-  tc.constraint_type = 'FOREIGN KEY' AND
-  kcu.table_schema NOT LIKE 'pg_%' AND
-  kcu.table_schema = $2 AND
-  kcu.table_name = $1
-GROUP BY
-  kcu.constraint_schema,
-  kcu.table_name,
-  rc.unique_constraint_schema,
-  rc.unique_constraint_name,
-  tc.constraint_name,
-  rc.update_rule,
-  rc.delete_rule;
-`;
-
-  const params = [
-    table,
-    schema,
-  ];
-
-  const data = await driverExecuteSingle(conn, { query: sql, params });
-
-  return data.rows.map((row) => ({
-    toTable: row.to_table,
-    toSchema: row.to_schema,
-    toColumn: row.to_column,
-    fromTable: row.from_table,
-    fromSchema: row.from_schema,
-    fromColumn: row.from_column,
-    constraintName: row.constraint_name,
-    onUpdate: row.update_rule,
-    onDelete: row.delete_rule
-  }));
-}
+  async createDatabase(databaseName: string, charset: string, _collation: string): Promise<void> {
+    const {isPostgres, number: versionAsInteger } = this.version;
 
-export async function getPrimaryKey(conn: HasPool, _database: string, table: string, schema: string): Promise<string | null> {
-  const keys = await getPrimaryKeys(conn, _database, table, schema)
-  return keys.length === 1 ? keys[0].columnName : null
-}
+    let sql = `create database ${wrapIdentifier(databaseName)} encoding ${wrapIdentifier(charset)}`;
 
-export async function getPrimaryKeys(conn: HasPool, _database: string, table: string, schema: string): Promise<PrimaryKeyColumn[]> {
-  const version = await getVersion(conn)
-  const tablename = PD.escapeString(tableName(table, schema), true)
-  const psqlQuery = `
-    SELECT
-      a.attname as column_name,
-      format_type(a.atttypid, a.atttypmod) AS data_type,
-      a.attnum as position
-    FROM   pg_index i
-    JOIN   pg_attribute a ON a.attrelid = i.indrelid
-                        AND a.attnum = ANY(i.indkey)
-    WHERE  i.indrelid = ${tablename}::regclass
-    AND    i.indisprimary
-    ORDER BY a.attnum
-  `
+    // postgres 9 seems to freak out if the charset isn't wrapped in single quotes and also requires the template https://www.postgresql.org/docs/9.3/sql-createdatabase.html
+    // later version don't seem to care
+    if (isPostgres && versionAsInteger < 100000) {
```

Comment:
Can we remove `isPostgres`? Other classes should override this if needed.

Reply:
Yup!